shading/lidar/Makefile

```diff
@@ -0,0 +1,385 @@
+endef
+
+
+# DMR 3.5 GKU
+
+build/dmr35/sk_dmr35_cut.tif: $(input_dmr35)
+	mkdir -p $(@D)
+	gdalwarp \
+		-of GTiff -ot Float32 -co COMPRESS=LZW -co TILED=YES -co BIGTIFF=YES \
+		-co NUM_THREADS=ALL_CPUS -wo NUM_THREADS=ALL_CPUS -multi \
+		-cutline $(make_data)/dmr35-cut.geojson -cl dmr35-cut -crop_to_cutline \
+		-r lanczos \
+		-tr 10 10 -tap -dstnodata -9999 \
+		-t_srs EPSG:8353 $^ $@
+	gdalinfo -stats $@
+
+build/dmr35/shading.tif: build/dmr35/sk_dmr35_cut.tif
+	$(call make_shading)
+
+# DMR 5.0 GKU
+
+build/dmr50/%/shading.tif: $(input_dmr50)/%.tif
+	$(call make_shading)
+	# zdroje z GKU nemaju vzdy dobre nastavenu projekciu, override
+	gdal_edit.py -a_srs EPSG:8353 $@
+	gdal_edit.py -a_srs EPSG:8353 $(@D)/cliffs.tif
+
+# polsko geoportal
+
+$(input_pl)/pl.vrt: $(input_pl)
+	cd $(input_pl) && ls *.tif > filelist.txt \
+		&& gdalbuildvrt -input_file_list filelist.txt pl.vrt
+
+# # resize old
+# build/pl/pl_jtsk.tif: $(input_pl)/pl.vrt
+# 	mkdir -p $(@D)
+# 	gdalwarp \
+# 		-of GTiff -ot Float32 -co COMPRESS=LZW -co TILED=YES -co BIGTIFF=YES \
+# 		-co NUM_THREADS=ALL_CPUS -wo NUM_THREADS=ALL_CPUS -multi \
+# 		-r cubic -order 3 \
+# 		-tr 1 1 -tap \
+# 		-t_srs EPSG:8353 $^ $@
+# 	gdalinfo -stats $@
+# 	gdaladdo \
+# 	-r lanczos --config BIGTIFF_OVERVIEW YES --config GDAL_NUM_THREADS ALL_CPUS \
+# 	-ro $@
+# build/pl/retiled/%.tif : build/pl/pl_jtsk.tif
+# 	mkdir -p $(@D)
+# 	gdalwarp \
+# 		-of GTiff -ot Float32 -co COMPRESS=LZW -co TILED=YES -co BIGTIFF=YES \
+# 		-co NUM_THREADS=ALL_CPUS -wo NUM_THREADS=ALL_CPUS -multi \
+# 		-cutline $(make_data)/pl_fat/$(basename $(@F)).gpkg -crop_to_cutline \
+# 		$^ $@
+# 	gdalinfo -stats $@
+# 	gdaladdo \
+# 	-r lanczos --config BIGTIFF_OVERVIEW YES --config GDAL_NUM_THREADS ALL_CPUS \
+# 	-ro $@
+
+# resize new in one step
+build/pl/retiled/%.tif: $(input_pl)/pl.vrt
+	mkdir -p $(@D)
+	gdalwarp \
+		-of GTiff -ot Float32 -co COMPRESS=LZW -co TILED=YES -co BIGTIFF=YES \
+		-co NUM_THREADS=ALL_CPUS -wo NUM_THREADS=ALL_CPUS -multi \
+		-cutline $(make_data)/pl_fat/$(basename $(@F)).gpkg -crop_to_cutline \
+		-r cubic -order 3 \
+		-tr 1 1 -tap \
+		-t_srs EPSG:8353 $^ $@
+	gdalinfo -stats $@
+	gdaladdo \
+	-r lanczos --config BIGTIFF_OVERVIEW YES --config GDAL_NUM_THREADS ALL_CPUS \
+	-ro $@
+
+build/pl/pl_jtsk_tiled.vrt: $(pl_dem_files)
+	gdalbuildvrt $@ build/pl/retiled/*.tif
+
+build/pl/%/shading.tif: build/pl/retiled/%.tif
+	$(call make_shading)
+	# no need to cut while merging, overlap not visible
+
+# SRTM
+
+$(input_srtm)/sk_srtm.vrt: $(input_srtm)
+	cd $(input_srtm) && gdalbuildvrt sk_srtm.vrt *.hgt
+
+build/srtm/sk_srtm_jtsk.tif: $(input_srtm)/sk_srtm.vrt
+	mkdir -p $(@D)
+	gdalwarp \
+		-of GTiff -ot Float32 -co COMPRESS=LZW -co TILED=YES -co BIGTIFF=YES \
+		-co NUM_THREADS=ALL_CPUS -wo NUM_THREADS=ALL_CPUS -multi \
+		-cutline $(make_data)/limit-slovensko-buffered.geojson -crop_to_cutline \
+		-r lanczos -order 3 \
+		-tr 20 20 -tap -dstnodata -9999 \

```

Comment:
tie `-tr` (plus pridat `-tap` alebo specifikovat `-ts`) treba optimalizovat na max zoom aby to bolo "pixel perfect". Vid skript na spodku https://github.com/FreemapSlovakia/freemap-operations/wiki/Generovanie-dla%C5%BEd%C3%ADc-Ortofotomozaika-SR

Reply:
Co mas na mysli s "pixel perfect"? Je to ako u [`-tap`](https://gdal.org/programs/gdalwarp.html#cmdoption-gdalwarp-tap), kde chces, aby hranica pixelu bola presne na rozhrani georeferencovanych jednotiek, takze napriklad tu presne kazdych 20 metrov? Ja som to pochopil tak, ze bud si vyratas hodnoty ako v tvojom skripte pre `-te` a `-ts` alebo staci dat `-tr` a `-tap` a gdal to zarovna a vysledny extent bude tiez zarovnany.

Reply:
pixel perfect - ze pixle geotiffu so shadingom budu presne sediet s pixlami tilov na nejakom zoome (na maximalnom zoome ktory ma este v DMR nejake detaily - pre DMR5.0 mi to vyslo na zoom 17). Skusim poslat obrazok ked deti pospia :-)

Reply:
Inak co sa tyka projekcii robim to takto:
1. najprv vsetko do JTSK03 (EPSG:8353)
2. v tom robim shading, ratam sklon, cliffs, atd. - pokial tomu spravne rozumiem, tu je jeden meter naozaj jeden meter (nie ako web mercator)
3. nakoniec vsetko do web mercator (EPSG:3857) s `-tr` a `-tap` pre zoom 17

Reply:
> pixel perfect - ze pixle geotiffu so shadingom budu presne sediet s pixlami tilov na nejakom zoome (na maximalnom zoome ktory ma este v DMR nejake detaily - pre DMR5.0 mi to vyslo na zoom 17). Skusim poslat obrazok ked deti pospia :-)

Dakujem. Tu ale este newarpujem do rozlisenia pre web mercator, ale len medzikrok do JTSK03. DMR5.0 a Polsko mam zarovnane na 1 meter, DMR3.5 na 10 metrov a SRTM na 20 metrov v JTSK03.

Az na konci warpujem na web:
Makefile riadok 173: `-tr 1.194328566968441 1.194328566968441 -tap`

Reply:
Ja pouzivam v QGISe EPSG:5514 a pride mi presny (podla toho je na freemape aj Orotofotomozaika SR a sedi to s WMS). Isiel som podla https://www.geoportal.sk/files/gz/s-jtsk_jtsk03_v_qgis.pdf. Dolezite je nainstalovat aj ten subor `Slovakia_JTSK03_to_JTSK.gsb` (a dat do aj do `/usr/share/proj/` aby ho pouzil aj gdal).

To, ze je jeden meter naozaj jeden meter, je podla mna jedno - ved aj tak v skripte pouzivas v `gdaldem` ten isty parameter `-z`. ktory upravuje vysku kopcov aby boli kompatibilne s mierkou plochy. ako som pouzil ja v EPSG:3857. A ja som si ho aj tak tusim len "vycucal z prsta" - proste metodou pokus omyl, ze kedy to bude pekne. _Mozno_ prave pre JTSK ho nie je nutne uviest.

Naviac ked potom spajas tie tienovania tak je nutne aby vsetky mali rovnaky pomer `-z` voci mierke plochy. Ak su generovane z roznych projekcii tak by som cakal rozne hodnoty `-z`. Alebo vsetko davas najprv do JTSK, aj tie SRTM1 od NASA? Potom je OK ze `-z` je rovnake. Len ci ma byt potom fakt 1.7 ;-). Ale aj tak...

...tym prvotnym warpovanim do JSTK na jeden meter sa zbytocne straca nejaka presnost. Proste warpovanim z nejakej mriezky na inu sa trochu presnosti strati (ak to povedzme nie je nasobok 2 zdrojovej mriezky). Uz len to, ze preco si zvolil 1 meter. Preco nie 1.1 alebo 0.9 :-). Takze podla mna bolo lepsie, ked sa rovno DMR dalo do "pixel zarovnanej" EPSG:3857 a z toho sa uz generovalo automaticky "pixel zarovnane" tienovanie.

Kludne ma oprav ak som nieco prehliadol :-).

Reply:
Dakujem za vysvetlenie, z tych projekcii som niekedy jelen... Zajtra na to kuknem a dam vediet.

Reply:
Hmm... Ten [navod na suradnicovy system JTSK03](https://www.geoportal.sk/files/gz/s-jtsk_jtsk03_v_qgis.pdf), ktory spominas, ma teraz verziu 3.0, a venuje sa hlavne EPSG::8353. EPSG::5514 sa tam spomina len kvoli transformacii do EPSG::8353. Ja som z toho pochopil, ze sa ma pouzivat ta novsia 8353.

Pre DMR je [tento navod na pracu s digitalnym modelom](https://www.geoportal.sk/files/zbgis/lls/navod-pracu-dmr-qgis.pdf) vo verzii 2.0. Tam odporucaju pouzivat  EPSG::8353 a pisu, ze starsie mali nespravne nastavene EPSG::5514 a treba to override. Ja si pamatam, ze som to cital kedysi pred rokom, vtedy bola asi starsia verzia, ktoru uz teraz neviem dohladat, a tam trebalo zas pouzivat nejaku custom projekciu ci co... 

Ale najvacsia haluz je, ze aj tak tvoj sposob lepsie sedi u ortofotomapy :-) Ja som si ju tiez robil do dlazdic pre oblast Nizkych Tatier, ked sme sli na trek. Pouzil som v QGISe EPSG:8353. Ked to porovnavam s WMS z Geoportalu - tvoja varianta sedi presne s WMS, moja je kusocek  posunuta na vychod asi 0.5 metra.

Je ten `Slovakia_JTSK03_to_JTSK.gsb` subor nutny aj na transformaciu z EPSG:8353 do WGS84 alebo web mercator? Lebo ked pouzivam EPSG:8353 - aspon u DMR5.0 mi to sedi celkom dobre na celom uzemi SR, ktore su uz k dispozicii, aj bez extra suborov. V tych navodoch som sa nedocital, ze by na pouzitie EPSG:8353 bolo treba v QGISe nieco extra nastavovat.

Reply:
Preco niektore oblasti transformujem 2x: 

Ked som prvy krat skusal robit utesy vo web mercator, narazil som na to, ze 1 meter u web mercator nie je naozaj jeden meter v skutocnosti a davalo mi to blbe vysledky pre sklon 45°. Nechcel som to nasobit nejakou bulharskou konstantou,  tak som sa rozhodol to cele robit v JTSK a v rozliseni, v akom je vacsina zdrojovych dat z DMR5.0 aj za cenu preprojekcie ostatnych. DMR5.0 je vacsinou 1 pixel = 1 meter zarovnane na cele metre - tak som si to chcel takto do VRT pospajat vsetko. SRTM a DMR3.5 aj tak nie je velmi presne, tam ma to warpovanie navyse nepali, Polsky LIDAR som si porovnaval - az tak to neutrpelo. 

Este som narazil na dalsi problem - ked som warpoval priamo DEM (float vyska) do web mercator, stavalo sa mi, ze sa to hnusne rozkockovalo - jedno, aky resampling som pouzil. Menej problemov robilo, ked som resamploval az shading (RGBA obrazok). Tak bolo vyhodnejsie spracovat shading v JTSK03 aby som aspon nemusel resamplovat DEM u DMR5.0.

Urcite najlepsie warpovat co najmenej a ked netreba ratat utesy - rovno do web mercator.

Reply:
Gdaldem `-z` :

Toto sa priznam sa som velmi neriesil. proste som skusil shading u vsetkych 4 zdrojov dat v JTSK03 (DMR5.0, DMR3.5, Polsky lidar, SRTM)  s Tvojimi parametrami. Poskladal som dokopy, porovnal. Nebolo velmi vidno rozdiel, ze by niektory bol tmavsy a iny svetlejsi, tak som to nechal tak.

Reply:
> Je ten Slovakia_JTSK03_to_JTSK.gsb subor nutny aj na transformaciu z EPSG:8353 do WGS84 alebo web mercator?

To neviem. Skus pozriet v tom PDF co som linkoval. Tam je aj nejaky screenshot kde QGIS upozornuje ze pri istej transformacii ten subor potrebuje. Ale nemam teraz viac casu na hranie sa s tym.

> Ked som prvy krat skusal robit utesy vo web mercator, narazil som na to, ze 1 meter u web mercator nie je naozaj jeden meter v skutocnosti a davalo mi to blbe vysledky pre sklon 45°. Nechcel som to nasobit nejakou bulharskou konstantou

No ved to cislo 1.7 v `-z` je asi prave ta bulharska konstanta. A neviem ci gdal v ramci projekcie rozumie co je "meter". Treba nastudovat :-). Tak ci tak by som skusil to leobit tym povodnym sposobom - iba jednym wrapovanim. Aby nebolo treba dalsiu bulharsku konstantu `-tr 1 1`.

> Este som narazil na dalsi problem - ked som warpoval priamo DEM (float vyska) do web mercator, stavalo sa mi, ze sa to hnusne rozkockovalo - jedno, aky resampling som pouzil.

Mne to pri floatoch nerobilo. Skus poslat detaily nech vyskusam u seba.

Reply:
Nemal som na mysli bulharsku konstantu u `gdaldem hillshade`. Ten `-z` je len koeficient ktorym sa nasobi vyska - shading kus viac stmavne. 
Mal som na mysli `gdaldem slope`, kde je nutne aby vstupna projekcia bola naozaj v metroch, ak chces mat vysledny sklon spravny. Myslim si, ze gdal je jedno, co je meter. Ale ak mas vysku v metroch a pixel nie je v metroch, slope bude skreslene.